src/os2datascanner/utils/optimiser/optimiser.py

```diff
@@ -0,0 +1,314 @@
+
+        *JSON to CustomContainer*"""
+
+        for k in container.components:
+            if isinstance(k, dict):
+                if k["type"] == "not":
+                    container.components = arr_switch(arr=container.components, deletion=k, insertion=-k["rule"])
+                else:
+                    new = CustomContainer(k["type"], k["components"], container)
+                    container.components = arr_switch(arr=container.components, deletion=k, insertion=new)
+                    self.containers.append(container.components[-1])
+                    yield container.components[-1]
+
+
+    def get_containers(self, main: 'CustomContainer'):
+        """Gets a list of all the containers.
+        Repeats itself with containify, but
+        otherwise the code wouldn't work."""
+        found = [main]
+        next_up: list[CustomContainer] = [main]
+        while next_up:
+            for cont in next_up:
+                for comp in cont.components:
+                    if isinstance(comp, dict):
+                        new = CustomContainer(comp["type"], comp["components"], cont)
+                        found.append(new)
+                        cont.components.remove(comp)
+                        cont.components.append(new)
+                        self.containers.append(new) # [new piece of code]
+                    elif isinstance(comp, CustomContainer):
+                        found.append(comp)
+                next_up.remove(cont)
+        return found
+
+
+    def check_empty(self, container:'CustomContainer'):
+        # Checks if container has no components i.e. AND(null) | OR(null)
+        if not container.components:
+            try:
+                container.parent.components.remove(container)
+                self.containers.remove(container)
+                self.found_redundancy = True
+            except ValueError:
+                pass
+            return True
+        return False
+
+
+    def check_useless(self, container:'CustomContainer'):
+        # Useless i.e. AND(1), OR(1), etc ...
+        if len(container.components) == 1:
+            try:
+                container.parent.components = arr_switch(container.parent.components, container, container.components[0])
+                self.containers.remove(container)
+                self.found_redundancy = True
+            except ValueError:
+                pass
+            return True
+        return False
+
+
+    def check_symbol_redundancy(self, container:'CustomContainer'):
+        # Checks for AND(n, -n) | OR(n, -n)
+        mem = []
+        for k in container.components:
+            mem.append(k)
+            if isinstance(k, int):
+                if -k in mem:
+                    container.parent.components.remove(container)
+                    self.containers.remove(container)
+                    self.found_redundancy = True
+                    return True
+        return False
+
+    def optimisation_cycle(self):
```

Comment:
This optimisation pass looks vaguely plausible, but you haven't implemented any unit tests for it, so it's hard to see what it actually does...

Reply:
Maybe i don't see the same quoted code as you do or you made a mistake, or do you mean in general ?